[Storage] Support Management Policy in Storage

blueww opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

<!-- Please add a brief description of the changes made in this PR -->

The design is in https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/57 .

There's an open issue, and the PR use the 2nd resolution to use object as input policy.

The Unit test is not ready ,will add later. Send the PR for product code review.

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

praries880 commented:
@blueww Also WRT the desktop failure, its because of help generation failures.
Here are the errors, do kindly fix them :

Target | Description
-- | --
Get-AzureRmStorageAccountManagementPolicy.md | Trailing yaml string found in description for parameter   StorageAccountName. Please move the trailing yaml string to a line by itself.
Remove-AzureRmStorageAccountManagementPolicy.md | Trailing yaml string found in description for parameter   StorageAccountName. Please move the trailing yaml string to a line by itself.
Set-AzureRmStorageAccountManagementPolicy.md | Trailing yaml string found in description for parameter   StorageAccountName. Please move the trailing yaml string to a line by itself.

praries880 commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/GetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,55 @@
+            HelpMessage = "Storage Account Name.")]
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
```
This should just be a "Name" you can add an alias to "StorageAccountName" if you want to.

blueww replied:
As the ManagementPolicy also has a property "Name", so don't want to ambiguous the Account name and Management policy Name , so use StorageAccountName  here.

Besides that, if use Name, pipe line the Policy object to remove-AzureRmStorageAccountManagementPolicy will fail, since the policy already has a property name, so the storage account name won't pipe to Name property.

markcowl replied:
+1 to Pary's comment - piping the name by proeprty name is now deprecated, the defautl piping mechanism should be the InputObject parameter set (see comments below)

markcowl replied:
Note that, for this cmdlet, it should also take piping by InputObject from Get-AzureRmStorageAccount

praries880 commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/RemoveAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,56 @@
+            HelpMessage = "Storage Account Name.")]
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
```
same comment as above

blueww replied:
same as above

markcowl replied:
This should not use piping by property name, but needs to support the piping mechanisms described here: https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/piping-in-powershell.md#azure-powershell-piping-scenarios

Specifically, you need to implement the InputObject and ResourceId parameter sets.  Piping by property name into the rgname + name parameter set is deprecated

praries880 commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/SetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,68 @@
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The Data Policy Rules.")]
+        [ValidateNotNullOrEmpty]
+        public object Policy { get; set; }
```
Is there a more specific type for Policy than "Object"

blueww replied:
Since we need to input a JObject generete in Json 6 here, but AzureRM.Storage is reference Json10, so we can't specify the correct input object type Other than Object.

praries880 commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/SetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,68 @@
+            HelpMessage = "Storage Account Name.")]
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
```
same comment as above

blueww replied:
same as above.
As the ManagementPolicy also has a property "Name", so don't want to ambiguous the Account name and Management policy Name , so use StorageAccountName here.

markcowl replied:
Same comments as previously

blueww commented:
As I discussed with Mark, set the JSON version in AzureRM.Storage can resolve the json serialize issue.
And update he Unit test. (Since the new API version only work on stage, so use stage location temperately.)

praries880 commented:
@blueww  The [linked design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/57) says that this is a public preview, as with the other public preview changes this should not target the current branch it is being merged into. Does the branch "AzureRM.Storage.Location" sound like a good branch to release this feature from?

blueww commented:
@praries880 

I have changed it to target to AzureAzureRM.Storage.Location.

This should be not merged before OAuth and Immutability policy release (target 5/18).
This feature target to release on 5/30 with preview module.

praries880 commented:
Static analysis errors  : 


c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Hyak.Common | 0 | 3000 | Missing assembly Newtonsoft.Json referenced   from Hyak.Common | Ensure that the assembly is included in the   Wix file or directory
-- | -- | -- | -- | -- | -- | -- | --
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.Azure.Commands.Common.Authentication.Abstractions | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.Azure.Commands.Common.Authentication.Abstractions | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.Azure.Commands.Management.Storage | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.Azure.Commands.Management.Storage | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.Azure.Commands.ResourceManager.Common | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.Azure.Commands.ResourceManager.Common | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.Azure.Management.Storage | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.Azure.Management.Storage | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.Rest.ClientRuntime.Azure | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.Rest.ClientRuntime.Azure | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.Rest.ClientRuntime | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.Rest.ClientRuntime | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 6.0.0.0 | Microsoft.WindowsAzure.Commands.Common | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.WindowsAzure.Commands.Common | Ensure that the assembly is included in the Wix file or directory
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage | Newtonsoft.Json | 10.0.0.0 | Microsoft.WindowsAzure.Storage | 0 | 3000 | Missing assembly Newtonsoft.Json referenced from   Microsoft.WindowsAzure.Storage | Ensure that the assembly is included in the Wix file or directory

blueww commented:
@praries880 

I have rebase the commit based on PR #6151, and the lasted code of branch "AzureRM.Storage.Location". Also updated the Unit test.

After rebase, the only static analisis error is following. As I discussed with Mark before, we should reference Json 6.0 in AzureRM.storage. And as AzureRM.profile contains both Json6.0 and Json 10.0, it will works fine. So I added this analisis failure to exception.

> "c:\code\PSH_Dev\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Storage","Newtonsoft.Json","6.0.0.0","Microsoft.Azure.Commands.Management.Storage","0","3000","Missing assembly Newtonsoft.Json referenced from Microsoft.Azure.Commands.Management.Storage","Ensure that the assembly is included in the Wix file or directory"

blueww commented:
@markcowl , @praries880 
The PR is target for the next release on AzureRm.Storage on 5/29. 
Would you please help to merge it if no other problem?

I would like to test RC when get it.

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/GetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,55 @@
+        [Parameter(
+            Position = 0,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
No more Value from pipeline by property name, see comments below.

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/RemoveAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,56 @@
+        [Parameter(
+            Position = 0,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
no more value from pipeline by propertyname for this parameter set

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/SetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,68 @@
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+            if (ShouldProcess(this.StorageAccountName, "Set Storage Account Data Policy"))
```
Please use a string resource for customer facing strings

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/help/Get-AzureRmStorageAccountManagementPolicy.md:
````diff
@@ -0,0 +1,93 @@
+### Example 1: Get the management policy of a Storage account.
+```
+PS C:\>Get-AzureRmStorageAccountManagementPolicy -ResourceGroupName "MyResourceGroup" -AccountName "mystorageaccount"
+```
````
Can you include sample output?

blueww replied:
added

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/help/Set-AzureRmStorageAccountManagementPolicy.md:
````diff
@@ -0,0 +1,180 @@
+    }]
+}'
+PS C:\>Set-AzureRmStorageAccountManagementPolicy -ResourceGroupName "MyResourceGroup" -AccountName "mystorageaccount" -Policy $policy
+```
````
Can you include sample output?

blueww replied:
added

blueww commented:
@markcowl 
Thanks for the review! I have resolved all your comments. Please help to look.
The target release is 6/3, and I will be OOF on 5/30-6/1, and 6/2-6/3 will be weekend.
So we are in the tight schedule. Please help to give the RC ASAP if no other comments, so I can have  the test ASAP.

BTW, it would be better if we can get the comments early to avoid the risk.
Thanks!

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/GetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Microsoft.Azure.Commands.Management.Storage
+{
+    [Cmdlet(VerbsCommon.Get, StorageAccountManagementPolicyNounStr), OutputType(typeof(PSManagementPolicy))]
```
Needs a DefaultParameterSetName [AccountNameParameterSet])

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/RemoveAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,102 @@
+
+namespace Microsoft.Azure.Commands.Management.Storage
+{
+    [Cmdlet(VerbsCommon.Remove, StorageAccountManagementPolicyNounStr, SupportsShouldProcess = true)]
```
Needs a DefaultParameterSetName (AccountNameParameterSet)

markcowl commented on src/ResourceManager/Storage/Commands.Management.Storage/StorageAccount/SetAzureStorageAccountManagementPolicy.cs:
```diff
@@ -0,0 +1,129 @@
+
+namespace Microsoft.Azure.Commands.Management.Storage
+{
+    [Cmdlet(VerbsCommon.Set, StorageAccountManagementPolicyNounStr, SupportsShouldProcess = true), OutputType(typeof(PSManagementPolicy))]
```
Needs a DefaultParameterSetName (AccountNamePolicyStringParameterSet)

blueww commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/904/
